### PR TITLE
[ML] partitions model definitions into chunks

### PR DESCRIPTION
**Pull request by benwtrent**
This paves the data layer way so that exceptionally large models are partitioned across multiple documents.

This change means that nodes before 7.8.0 will not be able to use trained inference models created on nodes on or after 7.8.0.

I chose the definition document limit to be 100. This *SHOULD* be plenty for any large model. One of the largest models that I have created so far had the following stats:
~314MB of inflated JSON, ~66MB when compressed, ~177MB of heap. 
With the chunking sizes of `16 * 1024 * 1024` its compressed string could be partitioned to 5 documents. 
Supporting models 20 times this size (compressed) seems adequate for now.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elastic/ml-docs where should I make note of the bwc limitations?

**Comment by szabosteve** [comment]
@benwtrent I think the best would be adding a `NOTE` or a `WARNING` to the PUT inference API docs: https://www.elastic.co/guide/en/elasticsearch/reference/master/put-inference.html
I'll also mention the bwc info in the limitations sections of the conceptual docs.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java [hunk 1]
```diff
@@ -82,6 +82,15 @@ protected void masterOperation(Task task,
                                    PutTrainedModelAction.Request request,
                                    ClusterState state,
                                    ActionListener<Response> listener) {
+        // 7.8.0 introduced splitting the model definition across multiple documents.
+        // This means that new models will not be usable on nodes that cannot handle multiple definition documents
+        if (state.nodes().getMinNodeVersion().before(Version.V_7_8_0)) {
+            listener.onFailure(ExceptionsHelper.badRequestException(
+                "Creating a new model that all nodes are at least version [{}]",
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java:89]:
```suggestion
                "Creating a new model requires that all nodes are at least version [{}]",
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java [hunk 2]
```diff
@@ -138,30 +143,40 @@ public void storeTrainedModel(TrainedModelConfig trainedModelConfig,
...
-                .setCompressionVersion(TrainedModelConfig.CURRENT_DEFINITION_COMPRESSION_VERSION)
-                .setDefinitionLength(compressedString.length())
-                .setTotalDefinitionLength(compressedString.length())
-                .build();
+            if (compressedString.length() > MAX_COMPRESSED_STRING_SIZE) {
+                listener.onFailure(
+                    ExceptionsHelper.badRequestException(
+                        "Unable to store model as compressed definition has length [{}] the limit is [{}]",
+                        compressedString.length(),
+                        MAX_COMPRESSED_STRING_SIZE));
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:154]: It should `return` here.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java [hunk 3]
```diff
@@ -677,13 +709,36 @@ private static QueryBuilder buildQueryIdExpressionQuery(String[] tokens, String
...
+        }
+        return results;
+    }
+
+    static List<String> chunkStringWithSize(String str, int chunkSize) {
+        List<String> subStrings = new ArrayList<>((int)Math.ceil(str.length()/(double)chunkSize));
+        for (int i = 0; i < str.length();i += chunkSize) {
+            subStrings.add(str.substring(i, Math.min(i + chunkSize, str.length())));
+        }
+        return subStrings;
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:741]: I am happy to merge this PR more-or-less as-is (see the two other minor comments I made).

However, in the long term if we're expecting to deal with models whose compressed definition is of the order of a gigabyte then I think more optimisations will be needed.  For example:

- Suppose this method is dealing with a `str` of length 1000000000
- That requires 2000000000 bytes to store as a UTF-16 `String`
- Then at the point this method returns every character has been copied into new strings but the original still exists so the memory usage at the point of return is over 4000000000 bytes

So we've gone from ~1GB of JSON model to temporarily using 4GB memory to manipulate it.

Since the compressed model is Base64, if it was stored as UTF-8 then that would be 1 byte per character.  And if this was wrapped in a `BytesReference` instead of a `String` then the returned `List<String>` could be a `List<BytesReference>` where the raw data was not duplicated but just referenced in slices by the list elements.

Basically, as we move from dealing with a few kilobytes of data to gigabytes we need to think much more about avoiding unnecessary duplication, most efficient character sets, etc.  But not in this PR, it can be done in a future one.

**Review by droberts195 — APPROVED** [review]
I saw two minor things and one bigger issue that can be deferred to a followup.

But I'm happy to merge this PR after just fixing the minor things.

**Comment by droberts195** [comment]
Also, I am not sure this PR needs to be labelled `>breaking`.  As I understand it, there is only a problem when putting models in the mixed version cluster, and we defend against this.  If a 7.8 cluster can still read models created in 7.7 then I don't think this PR needs to be documented as a breaking change.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java (hunk 3 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java:741]: @droberts195 yes, I realize this is an issue. My goal with this PR is to lay the groundwork, so that at least at the data layer (how things are stored/read in/from the index) we truly support more than one doc.
